File: apps/server/src/ports/driven/services/event-publisher.ts
// src/ports/driven/services/event-publisher.ts

import { Context, Data } from 'effect'
import type { Effect } from 'effect/Effect'

import type { DomainEvent } from '../../../domain'

// ============================================
// EVENT PUBLISHER ERROR
// ============================================

export class EventPublishError extends Data.TaggedError('EventPublishError')<{
  readonly event: DomainEvent
  readonly cause: unknown
}> {}

// ============================================
// EVENT PUBLISHER
// ============================================

export interface EventPublisherService {
  readonly publish: (event: DomainEvent) => Effect<void, EventPublishError>
}

export class EventPublisher extends Context.Tag('EventPublisher')<
  EventPublisher,
  EventPublisherService
>() {}

File: apps/server-java/src/main/java/com/maisonamane/port/driven/service/EventPublisher.java
package com.maisonamane.port.driven.service;

import com.maisonamane.port.driven.error.EventPublishError;
import io.vavr.control.Either;

/**
 * Interface for publishing domain events.
 * Maps Effect-TS Context.Tag("EventPublisher") to Java interface.
 */
public interface EventPublisher {

    /**
     * Publish a domain event.
     *
     * @param event the domain event to publish
     * @return Either containing an error on the left, or void (Unit) on the right
     */
    Either<EventPublishError, Void> publish(Object event);
}
